File: Python/Problem8.py
from collections import deque
from typing import List, Tuple
import copy
class Solution:
    def numberOfEnclaves(self, grid: List[List[int]]) -> int:
        n,m = len(grid), len(grid[0])
        vis = [[0]*m for _ in range(n)]
        q = deque()
        for i in range(n):
            for j in range(m):
                if i==0 or i==n-1 or j==0 or j==m-1:
                    if grid[i][j]==1:
                        q.append([i,j])
                        vis[i][j] = 1

        delrow = [-1, 0, 1, 0]
        delcol = [0, -1, 0, 1]
        while len(q):
            front = q.popleft()
            row = front[0]
            col = front[1]

            for i in range(4):
                newr = delrow[i] + row
                newc = delcol[i] + col
                if 0<=newr<n and 0<=newc<m and grid[newr][newc]==1 and not vis[newr][newc]:
                    vis[newr][newc] = 1
                    q.append([newr, newc])

        cnt = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j]==1 and not vis[i][j]:
                    cnt += 1

        return cnt

    def newGrid(self, grid: List[List[int]]) -> List[List[int]]:
        n,m = len(grid), len(grid[0])
        vis = [[0]*m for _ in range(n)]
        ans = copy.deepcopy(grid)
        q = deque()
        for i in range(n):
            for j in range(m):
                if i==0 or i==n-1 or j==0 or j==m-1:
                    if grid[i][j]==1:
                        q.append([i,j])
                        vis[i][j] = 1
                        ans[i][j] = 0

        delrow = [-1, 0, 1, 0]
        delcol = [0, -1, 0, 1]
        while len(q):
            front = q.popleft()
            row = front[0]
            col = front[1]

            for i in range(4):
                newr = delrow[i] + row
                newc = delcol[i] + col
                if 0<=newr<n and 0<=newc<m and grid[newr][newc]==1 and not vis[newr][newc]:
                    vis[newr][newc] = 1
                    q.append([newr, newc])
                    ans[newr][newc] = 0

        return ans

    def printGraph(self, grid: List[List[int]]) -> None:
        for row in grid:
            print(" ".join(map(str, row)))

if __name__ == "__main__":
    grid = [
        [0, 0, 0, 0],
        [1, 0, 1, 0],
        [0, 1, 1, 0],
        [0, 0, 0, 0]
    ]
    obj = Solution()
    print("old grid: ")
    obj.printGraph(grid)
    print("new grid: ")
    ans = obj.newGrid(grid)
    obj.printGraph(ans)
    cnt = obj.numberOfEnclaves(grid)
    print("The number of enclaves in the new grid: ", cnt)

File: JAVA/Problem8.java
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;

class Solution{
  class Pair{
    int row;
    int col;
    Pair(int row, int col){
      this.row = row;
      this.col = col;
    }
  }

  int numberOfEnclaves(int[][] grid){
    int n = grid.length;
    int m = grid[0].length;
    int[][] vis = new int[n][m];
    Queue<Pair> q = new LinkedList<Pair>();
    for(int i=0; i<n; i++){
      for(int j=0; j<m; j++){
        if(i==0 || j==0 || i==n-1 || j==m-1){
          if(grid[i][j]==1){
            vis[i][j] = 1;
            q.add(new Pair(i, j));
          }
        }
      }
    }
    int[] delrow = {-1, 0, 1, 0};
    int[] delcol = {0, -1, 0, 1};
    while(!q.isEmpty()){
      int row = q.peek().row;
      int col = q.peek().col;
      q.remove();

      for(int i=0; i<4; i++){
        int newr = delrow[i] + row;
        int newc = delcol[i] + col;

        if(newr>=0 && newr<n && newc>=0 && newc<m && grid[newr][newc]==1 && vis[newr][newc]==0){
          vis[newr][newc] = 1;
          q.add(new Pair(newr, newc));
        }
      }
    }

    int cnt = 0;
    for(int i=0; i<n; i++){
      for(int j=0; j<m; j++){
        if(grid[i][j]==1 && vis[i][j]==0){
          cnt++;
        }
      }
    }
    return cnt;
  }

  int[][] newGrid(int[][] grid){
    int n = grid.length;
    int m = grid[0].length;
    int[][] vis = new int[n][m];
    Queue<Pair> q = new LinkedList<Pair>();
    int[][] ans = new int[n][m];

    for (int i = 0; i < grid.length; i++) {
        for (int j = 0; j < grid[i].length; j++) {
            ans[i][j] = grid[i][j];
        }
    }

    for(int i=0; i<n; i++){
      for(int j=0; j<m; j++){
        if(i==0 || j==0 || i==n-1 || j==m-1){
          if(grid[i][j]==1){
            vis[i][j] = 1;
            q.add(new Pair(i, j));
            ans[i][j] = 0;
          }
        }
      }
    }
    int[] delrow = {-1, 0, 1, 0};
    int[] delcol = {0, -1, 0, 1};
    while(!q.isEmpty()){
      int row = q.peek().row;
      int col = q.peek().col;
      q.remove();

      for(int i=0; i<4; i++){
        int newr = delrow[i] + row;
        int newc = delcol[i] + col;

        if(newr>=0 && newr<n && newc>=0 && newc<m && grid[newr][newc]==1 && vis[newr][newc]==0){
          vis[newr][newc] = 1;
          q.add(new Pair(newr, newc));
          ans[newr][newc] = 0;
        }
      }
    }

    return ans;
  }

  void printGraph(int[][] grid){
    for(int[] row: grid){
      for(int val: row){
        System.out.print(val + " ");
      }
      System.out.println();
    }
  }

  public static void main(String[] args){
    int[][] grid = {{0, 0, 0, 1},
            {0, 1, 1, 0},
            {0, 1, 1, 0},
            {0, 0, 0, 1},
            {0, 1, 1, 0}};
    System.out.println("old grid: ");
    Solution obj = new Solution();
    obj.printGraph(grid);
    int[][] ans = obj.newGrid(grid);
    System.out.println("new grid: ");
    obj.printGraph(ans);
    int cnt = obj.numberOfEnclaves(grid);
    System.out.println("Number of enclaves in new grid: " + cnt);
  }
}
